test/test_sparse.py

```diff
@@ -143,6 +143,79 @@ def f5():
             # Check warn-once:
             self.assertEqual(len(cm.warnings), 1)
 
+    def test_gradcheck_check_sparse_nnz(self):
+        """Tests for deprecated check_sparse_nnz keyword argument of gradcheck.
+
+        Suggested deprecation steps:
+        2.1: Specification of check_sparse_nnz triggers a warning.
+        2.2: Specification of check_sparse_nnz triggers a warning and
+             an exception. Update gradcheck and this test accordingly.
+        2.3: Specification of check_sparse_nnz triggers an
+             exception. Remove all check_sparse_nnz usages from
+             gradcheck and delete this test.
+
+        """
+        def fn(x, masked_grad):
+            return x.to_dense(masked_grad=masked_grad)
+
+        def test(x, masked_grad, masked, check_sparse_nnz):
+            x = x.detach().clone().requires_grad_()
+            r = torch.autograd.gradcheck(fn, (x, masked_grad), masked=masked, check_sparse_nnz=check_sparse_nnz)
+            self.assertTrue(r)
+
+        x = torch.tensor([[0, 2], [3, 4]], dtype=torch.float64).to_sparse()
+
+        # Supported calls:
```

Comment:
Thanks for the detailed testing. But just to make it clear that the cases covered are collectively exhaustive, I'd prefer if we refactored this to all run under a single loop of all combinations. This would also allow us to logically define the rules for when certain things should pass, fail, warn, etc. 

Super simplified example:
```python
for masked_grad, masked, check_sparse_nnz in itertools.product(*[(True, False)]*3):
    if check_sparse_nnz != masked:
         with self.assertRaisesRegex(ValueError, "Expected specified check_sparse_nnz"):
              test(x, masked_grad, masked, check_sparse_nnz)
              
    if check_sparse_nnz is not None:
         with self.assertWarns(UserWarning, check_sparse_nnz will be deprecated):
             test(x, masked_grad, masked, check_sparse_nnz)
```